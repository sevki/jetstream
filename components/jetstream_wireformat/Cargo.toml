[package]
name = "jetstream_wireformat"
version = "7.1.0"
edition = { workspace = true }
description = { workspace = true }
license = { workspace = true }
repository = { workspace = true }
readme = { workspace = true }
documentation.workspace = true
[lib]
bench = false

[features]
default = ["async"]
async = ["dep:tokio", "dep:futures"]

[dependencies]
bytes = { version = "1.7.2", features = ["serde"] }
jetstream_macros = { version = "7.1.0", path = "../jetstream_macros" }

libc = "0.2.159"
paste = "1.0.15"
zerocopy = "0.8.3"
tokio = { version = "1.40.0", features = ["full"], optional = true }
okstd = { version = "0.1.10" }
futures = { version = "0.3.31", optional = true }
tokio-util = { version = "0.7.12", features = ["full", "io", "time"] }
time = "0.3.36"
