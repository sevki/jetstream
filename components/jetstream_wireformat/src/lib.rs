#![doc(
    html_logo_url = "https://raw.githubusercontent.com/sevki/jetstream/main/logo/JetStream.png"
)]
#![doc(
    html_favicon_url = "https://raw.githubusercontent.com/sevki/jetstream/main/logo/JetStream.png"
)]
#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
// Copyright (c) 2024, Sevki <s@sevki.io>
// Copyright 2018 The ChromiumOS Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
pub use jetstream_derive::JetStreamWireFormat;

use bytes::Buf;
use std::fmt;
use std::io;
use std::io::ErrorKind;
use std::io::Read;
use std::io::Write;
use std::mem;
use std::ops::Deref;
use std::ops::DerefMut;
use std::string::String;
use std::vec::Vec;
use zerocopy::{IntoBytes, LittleEndian};
pub mod wire_format_extensions;

/// A type that can be encoded on the wire using the 9P protocol.
pub trait WireFormat: std::marker::Sized + Send {
    /// Returns the number of bytes necessary to fully encode `self`.
    fn byte_size(&self) -> u32;

    /// Encodes `self` into `writer`.
    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()>;

    /// Decodes `Self` from `reader`.
    fn decode<R: Read>(reader: &mut R) -> io::Result<Self>;
}

// This doesn't really _need_ to be a macro but unfortunately there is no trait bound to
// express "can be casted to another type", which means we can't write `T as u8` in a trait
// based implementation.  So instead we have this macro, which is implemented for all the
// stable unsigned types with the added benefit of not being implemented for the signed
// types which are not allowed by the protocol.
macro_rules! uint_wire_format_impl {
    ($Ty:ty) => {
        impl WireFormat for $Ty {
            fn byte_size(&self) -> u32 {
                mem::size_of::<$Ty>() as u32
            }

            fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
                writer.write_all(&self.as_bytes())
            }

            fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
                let mut buf = [0; mem::size_of::<$Ty>()];
                reader.read_exact(&mut buf)?;
                paste::expr! {
                    let num: zerocopy::[<$Ty:snake:upper>]<LittleEndian> =  zerocopy::byteorder::[<$Ty:snake:upper>]::from_bytes(buf);
                    Ok(num.get())
                }
            }
        }
    };
}
// unsigned integers
uint_wire_format_impl!(u16);
uint_wire_format_impl!(u32);
uint_wire_format_impl!(u64);
uint_wire_format_impl!(u128);
// signed integers
uint_wire_format_impl!(i16);
uint_wire_format_impl!(i32);
uint_wire_format_impl!(i64);
uint_wire_format_impl!(i128);

impl WireFormat for u8 {
    fn byte_size(&self) -> u32 {
        1
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        writer.write_all(&[*self])
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let mut byte = [0u8; 1];
        reader.read_exact(&mut byte)?;
        Ok(byte[0])
    }
}

// The 9P protocol requires that strings are UTF-8 encoded.  The wire format is a u16
// count |N|, encoded in little endian, followed by |N| bytes of UTF-8 data.
impl WireFormat for String {
    fn byte_size(&self) -> u32 {
        (mem::size_of::<u16>() + self.len()) as u32
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        if self.len() > u16::MAX as usize {
            return Err(io::Error::new(
                ErrorKind::InvalidInput,
                "string is too long",
            ));
        }

        (self.len() as u16).encode(writer)?;
        writer.write_all(self.as_bytes())
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let len: u16 = WireFormat::decode(reader)?;
        let mut result = String::with_capacity(len as usize);
        reader.take(len as u64).read_to_string(&mut result)?;
        Ok(result)
    }
}

// The wire format for repeated types is similar to that of strings: a little endian
// encoded u16 |N|, followed by |N| instances of the given type.
impl<T: WireFormat> WireFormat for Vec<T> {
    fn byte_size(&self) -> u32 {
        mem::size_of::<u16>() as u32
            + self.iter().map(|elem| elem.byte_size()).sum::<u32>()
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        if self.len() > u16::MAX as usize {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "too many elements in vector",
            ));
        }

        (self.len() as u16).encode(writer)?;
        for elem in self {
            elem.encode(writer)?;
        }

        Ok(())
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let len: u16 = WireFormat::decode(reader)?;
        let mut result = Vec::with_capacity(len as usize);

        for _ in 0..len {
            result.push(WireFormat::decode(reader)?);
        }

        Ok(result)
    }
}

/// A type that encodes an arbitrary number of bytes of data.  Typically used for Rread
/// Twrite messages.  This differs from a `Vec<u8>` in that it encodes the number of bytes
/// using a `u32` instead of a `u16`.
#[derive(PartialEq, Eq, Clone)]
#[repr(transparent)]
pub struct Data(pub Vec<u8>);

// The maximum length of a data buffer that we support.  In practice the server's max message
// size should prevent us from reading too much data so this check is mainly to ensure a
// malicious client cannot trick us into allocating massive amounts of memory.
const MAX_DATA_LENGTH: u32 = 32 * 1024 * 1024;

impl fmt::Debug for Data {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        // There may be a lot of data and we don't want to spew it all out in a trace.  Instead
        // just print out the number of bytes in the buffer.
        write!(f, "Data({} bytes)", self.len())
    }
}

// Implement Deref and DerefMut so that we don't have to use self.0 everywhere.
impl Deref for Data {
    type Target = Vec<u8>;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl DerefMut for Data {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<T> WireFormat for Option<T>
where
    T: WireFormat,
{
    fn byte_size(&self) -> u32 {
        1 + match self {
            None => 0,
            Some(value) => value.byte_size(),
        }
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        match self {
            None => WireFormat::encode(&0u8, writer),
            Some(value) => {
                WireFormat::encode(&1u8, writer)?;
                WireFormat::encode(value, writer)
            }
        }
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let tag: u8 = WireFormat::decode(reader)?;
        match tag {
            0 => Ok(None),
            1 => Ok(Some(WireFormat::decode(reader)?)),
            _ => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Invalid Option tag: {}", tag),
            )),
        }
    }
}

// Same as Vec<u8> except that it encodes the length as a u32 instead of a u16.
impl WireFormat for Data {
    fn byte_size(&self) -> u32 {
        mem::size_of::<u32>() as u32
            + self.iter().map(|elem| elem.byte_size()).sum::<u32>()
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        if self.len() > u32::MAX as usize {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                "data is too large",
            ));
        }
        (self.len() as u32).encode(writer)?;
        writer.write_all(self)
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let len: u32 = WireFormat::decode(reader)?;
        if len > MAX_DATA_LENGTH {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                format!("data length ({} bytes) is too large", len),
            ));
        }

        let mut buf = Vec::with_capacity(len as usize);
        reader.take(len as u64).read_to_end(&mut buf)?;

        if buf.len() == len as usize {
            Ok(Data(buf))
        } else {
            Err(io::Error::new(
                std::io::ErrorKind::UnexpectedEof,
                format!(
                    "unexpected end of data: want: {} bytes, got: {} bytes",
                    len,
                    buf.len()
                ),
            ))
        }
    }
}

impl WireFormat for () {
    fn byte_size(&self) -> u32 {
        0
    }

    fn encode<W: Write>(&self, _writer: &mut W) -> io::Result<()> {
        Ok(())
    }

    fn decode<R: Read>(_reader: &mut R) -> io::Result<Self> {
        Ok(())
    }
}

impl WireFormat for bool {
    fn byte_size(&self) -> u32 {
        1
    }

    fn encode<W: Write>(&self, writer: &mut W) -> io::Result<()> {
        writer.write_all(&[*self as u8])
    }

    fn decode<R: Read>(reader: &mut R) -> io::Result<Self> {
        let mut byte = [0u8; 1];
        reader.read_exact(&mut byte)?;
        match byte[0] {
            0 => Ok(false),
            1 => Ok(true),
            _ => Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "invalid byte for bool",
            )),
        }
    }
}

impl io::Read for Data {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        self.0.reader().read(buf)
    }
}
