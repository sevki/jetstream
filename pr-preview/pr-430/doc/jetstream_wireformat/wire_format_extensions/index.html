<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `wire_format_extensions` mod in crate `jetstream_wireformat`."><title>jetstream_wireformat::wire_format_extensions - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="jetstream_wireformat" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (844264add 2025-10-14)" data-channel="nightly" data-search-js="search-8d3311b9.js" data-stringdex-js="stringdex-828709d0.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="icon" href="https://raw.githubusercontent.com/sevki/jetstream/main/logo/JetStream.png"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module wire_format_extensions</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../jetstream_wireformat/index.html"><img src="https://raw.githubusercontent.com/sevki/jetstream/main/logo/JetStream.png" alt="logo"></a><h2><a href="../../jetstream_wireformat/index.html">jetstream_<wbr>wireformat</a><span class="version">9.3.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module wire_<wbr>format_<wbr>extensions</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate jetstream_<wbr>wireformat</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">jetstream_wireformat</a></div><h1>Module <span>wire_<wbr>format_<wbr>extensions</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/jetstream_wireformat/wire_format_extensions.rs.html#4-244">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="tokio/index.html" title="mod jetstream_wireformat::wire_format_extensions::tokio">tokio</a></dt></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.AsyncWireFormat.html" title="trait jetstream_wireformat::wire_format_extensions::AsyncWireFormat">Async<wbr>Wire<wbr>Format</a></dt><dt><a class="trait" href="trait.ConvertWireFormat.html" title="trait jetstream_wireformat::wire_format_extensions::ConvertWireFormat">Convert<wbr>Wire<wbr>Format</a></dt><dd>A trait for converting types to and from a wire format.</dd></dl></section></div></main></body></html>