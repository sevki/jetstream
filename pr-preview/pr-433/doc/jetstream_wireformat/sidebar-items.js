window.SIDEBAR_ITEMS = {"derive":["JetStreamWireFormat"],"mod":["wire_format_extensions"],"struct":["Data","P9String","Wrapped"],"trait":["WireFormat"]};