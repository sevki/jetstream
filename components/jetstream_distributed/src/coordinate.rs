//! Vivaldi coordinate system implementation, based on the paper:
//! [Vivaldi: A Decentralized Network Coordinate System](https://pdos.csail.mit.edu/papers/vivaldi:sigcomm/paper.pdf)
//! by Frank Dabek, Russ Cox, Frans Kaashoek, Robert Morris
//! This implementation is based on the Go implementation by Hashicorp's [Serf](https://github.com/hashicorp/serf/)
//! ```mermaid
//! graph TB
//!    subgraph "Vivaldi Coordinate System"
//!        Node1[Node A] --> Vec1[Vector Components]
//!        Node1 --> H1[Height]
//!        Node1 --> Adj1[Adjustment]
//!
//!        Node2[Node B] --> Vec2[Vector Components]
//!        Node2 --> H2[Height]
//!        Node2 --> Adj2[Adjustment]
//!
//!        Vec1 & Vec2 --> ED[Euclidean Distance]
//!        H1 & H2 --> HD[Height Distance]
//!        Adj1 & Adj2 --> ADJ[Adjustment Factor]
//!
//!        ED & HD & ADJ --> TD[Total Distance]
//!    end
//!
//!    classDef component fill:#f9f,stroke:#333
//!    classDef calculation fill:#bbf,stroke:#333
//!
//!    class Vec1,Vec2,H1,H2,Adj1,Adj2 component
//!    class ED,HD,ADJ,TD calculation
//!
use jetstream_wireformat::JetStreamWireFormat;
use rand::Rng;
use std::time::Duration;

// Constants
const SECONDS_TO_NANOSECONDS: f64 = 1.0e9;
const ZERO_THRESHOLD: f64 = 1.0e-6;

/// Configuration for the Vivaldi coordinate system
#[derive(Debug, Clone, PartialEq, JetStreamWireFormat)]

pub struct Config {
    /// The dimensionality of the coordinate system
    pub dimensionality: usize,
    /// Maximum error value and default for new coordinates
    pub vivaldi_error_max: f64,
    /// Controls maximum impact of observations on confidence
    pub vivaldi_ce: f64,
    /// Controls maximum impact of observations on coordinates
    pub vivaldi_cc: f64,
    /// Number of samples for adjustment window (0 disables feature)
    pub adjustment_window_size: usize,
    /// Minimum height value (must be positive)
    pub height_min: f64,
    /// Maximum samples retained per node for latency filtering
    pub latency_filter_size: usize,
    /// Tuning factor for gravity effect
    pub gravity_rho: f64,
}

impl Default for Config {
    fn default() -> Self {
        Config {
            dimensionality: 8,
            vivaldi_error_max: 1.5,
            vivaldi_ce: 0.25,
            vivaldi_cc: 0.25,
            adjustment_window_size: 20,
            height_min: 10.0e-6,
            latency_filter_size: 3,
            gravity_rho: 150.0,
        }
    }
}

/// A Vivaldi coordinate
#[derive(Debug, Clone, JetStreamWireFormat)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Coordinate {
    /// Euclidean portion of the coordinate (in seconds)
    vec: Vec<f64>,
    /// Confidence in the coordinate (dimensionless)
    error: f64,
    /// Distance offset based on observations (in seconds)
    adjustment: f64,
    /// Distance offset for non-Euclidean effects (in seconds)
    height: f64,
}

/// Errors that can occur when working with coordinates
#[derive(Debug, thiserror::Error)]
pub enum CoordinateError {
    /// The coordinate dimensionality does not match
    #[error("coordinate dimensionality does not match")]
    DimensionalityConflict,
    /// The coordinate contains invalid values
    #[error("coordinate contains invalid values")]
    InvalidValues,
    /// The round trip time is not in a valid range
    #[error("round trip time not in valid range")]
    InvalidRtt,
}

impl Coordinate {
    /// Creates a new coordinate at the origin
    pub fn new(config: &Config) -> Self {
        Self {
            vec: vec![0.0; config.dimensionality],
            error: config.vivaldi_error_max,
            adjustment: 0.0,
            height: config.height_min,
        }
    }

    /// Checks if all components of the coordinate are valid
    pub fn is_valid(&self) -> bool {
        self.vec.iter().all(|&x| Self::component_is_valid(x))
            && Self::component_is_valid(self.error)
            && Self::component_is_valid(self.adjustment)
            && Self::component_is_valid(self.height)
    }

    fn component_is_valid(f: f64) -> bool {
        !f.is_infinite() && !f.is_nan()
    }

    /// Checks dimensional compatibility with another coordinate
    pub fn is_compatible_with(&self, other: &Coordinate) -> bool {
        self.vec.len() == other.vec.len()
    }

    /// Returns the distance between this coordinate and another
    pub fn distance_to(
        &self,
        other: &Coordinate,
    ) -> Result<Duration, CoordinateError> {
        if !self.is_compatible_with(other) {
            return Err(CoordinateError::DimensionalityConflict);
        }

        let dist = self.raw_distance_to(other);
        let adjusted_dist = dist + self.adjustment + other.adjustment;
        let final_dist = if adjusted_dist > 0.0 {
            adjusted_dist
        } else {
            dist
        };

        Ok(Duration::from_nanos(
            (final_dist * SECONDS_TO_NANOSECONDS) as u64,
        ))
    }

    /// Calculates raw Vivaldi distance without adjustments
    fn raw_distance_to(&self, other: &Coordinate) -> f64 {
        let euclidean_dist = self
            .vec
            .iter()
            .zip(other.vec.iter())
            .map(|(&a, &b)| (a - b).powi(2))
            .sum::<f64>()
            .sqrt();

        euclidean_dist + self.height + other.height
    }

    /// Applies force from another coordinate's direction
    pub fn apply_force(
        &self,
        config: &Config,
        force: f64,
        other: &Coordinate,
    ) -> Result<Coordinate, CoordinateError> {
        if !self.is_compatible_with(other) {
            return Err(CoordinateError::DimensionalityConflict);
        }

        let (unit, mag) = self.unit_vector_at(other);
        let mut new_coord = self.clone();

        // Update vector components
        (0..self.vec.len()).for_each(|i| {
            new_coord.vec[i] += unit[i] * force;
        });

        // Update height if points aren't too close
        if mag > ZERO_THRESHOLD {
            new_coord.height = (new_coord.height + other.height) * force / mag
                + new_coord.height;
            new_coord.height = new_coord.height.max(config.height_min);
        }

        Ok(new_coord)
    }

    /// Calculates unit vector pointing from this coordinate to another
    fn unit_vector_at(&self, other: &Coordinate) -> (Vec<f64>, f64) {
        let mut diff: Vec<f64> = self
            .vec
            .iter()
            .zip(other.vec.iter())
            .map(|(&a, &b)| a - b)
            .collect();

        let mag = (diff.iter().map(|&x| x * x).sum::<f64>()).sqrt();

        // If points aren't too close, normalize the vector
        if mag > ZERO_THRESHOLD {
            for d in diff.iter_mut() {
                *d /= mag;
            }
            return (diff, mag);
        }

        // Generate random unit vector if points are too close
        // SAFETY: this is not used for cryptographic purposes
        // so it's fine to use the default RNG
        let mut rng = rand::thread_rng();
        diff = diff.iter().map(|_| rng.gen::<f64>() - 0.5).collect();
        let mag = (diff.iter().map(|&x| x * x).sum::<f64>()).sqrt();

        if mag > ZERO_THRESHOLD {
            for d in diff.iter_mut() {
                *d /= mag;
            }
            return (diff, 0.0);
        }

        // Last resort: unit vector along first dimension
        let mut unit = vec![0.0; self.vec.len()];
        unit[0] = 1.0;
        (unit, 0.0)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use rand::seq::SliceRandom;

    use super::*;

    fn _verify_equal_vectors(v1: &[f64], v2: &[f64], epsilon: f64) {
        assert_eq!(v1.len(), v2.len());
        for (a, b) in v1.iter().zip(v2.iter()) {
            assert!((a - b).abs() < epsilon, "Expected {a} to equal {b}");
        }
    }

    #[test]
    fn test_new_coordinate() {
        let config = Config::default();
        let coord = Coordinate::new(&config);
        assert_eq!(coord.vec.len(), config.dimensionality);
        assert_eq!(coord.error, config.vivaldi_error_max);
        assert_eq!(coord.height, config.height_min);
    }

    #[test]
    fn test_is_valid() {
        let config = Config::default();
        let mut coord = Coordinate::new(&config);
        assert!(coord.is_valid());

        coord.vec[0] = f64::NAN;
        assert!(!coord.is_valid());

        coord.vec[0] = f64::INFINITY;
        assert!(!coord.is_valid());
    }

    #[test]
    fn test_incompatible_dimensions() {
        let mut config = Config {
            dimensionality: 3,
            ..Default::default()
        };

        let coord1 = Coordinate::new(&config);

        config.dimensionality = 2;
        let coord2 = Coordinate::new(&config);

        assert!(!coord1.is_compatible_with(&coord2));
        assert!(coord1.distance_to(&coord2).is_err());
    }
    #[test]
    fn test_distance_calculation() {
        let config = Config::default();
        let mut coord1 = Coordinate::new(&config);
        let mut coord2 = Coordinate::new(&config);

        // Set first 3 dimensions and pad rest with zeros
        coord1.vec = vec![-0.5, 1.3, 2.4];
        coord2.vec = vec![1.2, -2.3, 3.4];

        while coord1.vec.len() < config.dimensionality {
            coord1.vec.push(0.0);
            coord2.vec.push(0.0);
        }

        // Calculate euclidean distance
        let euclidean_distance = 4.104875150354758;

        // Expected total distance includes heights
        let expected_distance =
            euclidean_distance + coord1.height + coord2.height;

        let dist = coord1.distance_to(&coord2).unwrap();
        let got = dist.as_secs_f64();
        let diff = (got - expected_distance).abs();

        println!("Distance calculation:");
        println!("Got:      {}", got);
        println!("Expected: {}", expected_distance);
        println!("Diff:     {}", diff);

        assert!(
            diff < 1e-6,
            "Distance calculation failed:\nGot {}\nWanted {}\nDiff {}",
            got,
            expected_distance,
            diff
        );

        // Additional verification tests
        let mut height_free_coord1 = coord1.clone();
        let mut height_free_coord2 = coord2.clone();

        // Test with zero heights
        height_free_coord1.height = 0.0;
        height_free_coord2.height = 0.0;

        let pure_euclidean =
            height_free_coord1.distance_to(&height_free_coord2).unwrap();
        assert!(
            (pure_euclidean.as_secs_f64() - euclidean_distance).abs() < 1e-6,
            "Pure Euclidean distance failed:\nGot {}\nWanted {}",
            pure_euclidean.as_secs_f64(),
            euclidean_distance
        );
    }

    // Add test for pure height calculation
    #[test]
    fn test_height_contribution() {
        let config = Config::default();
        let coord1 = Coordinate::new(&config);
        let coord2 = Coordinate::new(&config);

        // With default config, both coords should have height_min
        let expected_height_contribution = config.height_min * 2.0;
        let dist = coord1.distance_to(&coord2).unwrap();

        assert!(
            (dist.as_secs_f64() - expected_height_contribution).abs() < 1e-6,
            "Height-only distance failed:\nGot {}\nWanted {}",
            dist.as_secs_f64(),
            expected_height_contribution
        );
    }

    // Add test for adjustment contribution
    #[test]
    fn test_adjustment_contribution() {
        let config = Config::default();
        let mut coord1 = Coordinate::new(&config);
        let mut coord2 = Coordinate::new(&config);

        coord1.adjustment = 0.1;
        coord2.adjustment = 0.2;

        let base_dist = coord1.height + coord2.height; // Just heights, vectors are 0
        let expected_dist = base_dist + 0.3; // 0.3 is sum of adjustments

        let dist = coord1.distance_to(&coord2).unwrap();

        assert!(
            (dist.as_secs_f64() - expected_dist).abs() < 1e-6,
            "Adjustment distance failed:\nGot {}\nWanted {}",
            dist.as_secs_f64(),
            expected_dist
        );
    }

    // Generate a random map of coordinates with 20 different nodes.
    // This is a simple test to verify that the map is generated correctly.
    #[test]
    fn test_map_generation() {
        let config = Config::default();
        let mut coordinates = vec![Coordinate::new(&config); 20];
        let mut node_map: HashMap<String, Coordinate> = HashMap::new();

        // crates a rand that's deterministic with a seed
        let mut rng = rand::rngs::mock::StepRng::new(2, 10);

        // apply force to each coordinate in relation to another random coordinate
        Iterator::for_each(0..coordinates.len(), |i| {
            let other = coordinates.choose(&mut rng).unwrap();

            let force: f64 = rng.gen_range(0.0..2.0);
            coordinates[i] =
                coordinates[i].apply_force(&config, force, other).unwrap();

            // Insert into the map with "node_{i}" as the key
            node_map.insert(format!("node_{}", i), coordinates[i].clone());
        });
        // sort the map so test doesn't fail due to different ordering
        let mut node_map: Vec<_> = node_map.into_iter().collect();
        node_map.sort_by(|a, b| a.0.cmp(&b.0));
        #[cfg(feature = "serde")]
        {
            insta::assert_json_snapshot!(node_map);
        }
    }
}
