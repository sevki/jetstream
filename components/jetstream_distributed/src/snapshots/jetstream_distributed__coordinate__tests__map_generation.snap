---
source: components/jetstream_distributed/src/coordinate.rs
expression: node_map
---
[
  [
    "node_0",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_1",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_10",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_11",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_12",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_13",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_14",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_15",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_16",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_17",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_18",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_19",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_2",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_3",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_4",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_5",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_6",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_7",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_8",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ],
  [
    "node_9",
    {
      "vec": [
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0,
        0.0
      ],
      "error": 1.5,
      "adjustment": 0.0,
      "height": 0.00001
    }
  ]
]
